devito/arguments.py

```diff
@@ -0,0 +1,183 @@
+from cached_property import cached_property
+import cgen as c
+import numpy as np
+
```

Comment:
I feel like some of the type/variable names below don't follow a convention? 
E.g., Arg vs Argument, Prov vs Provider